File: java/src/main/java/com/aaronicsubstances/shrewd/evolver/LogMessageTemplate.java
package com.aaronicsubstances.shrewd.evolver;

import com.aaronicsubstances.shrewd.evolver.LogMessageTemplateParser.PartDescriptor;

import java.lang.reflect.Method;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;

public abstract class LogMessageTemplate {
    
    public static class Unstructured {
        private final String formatString;
        private final List<Object> formatArgs;

        public Unstructured(String formatString, List<Object> formatArgs) {
            this.formatString = formatString;
            this.formatArgs = formatArgs;
        }

        public String getFormatString() {
            return formatString;
        }

        public List<Object> getFormatArgs() {
            return formatArgs;
        }
    }

    public class Structured {
        private final Object treeDataSlice;

        public Structured(Object treeDataSlice) {
            this.treeDataSlice = treeDataSlice;
        }

        @Override
        public String toString() {
            return serializeData(treeDataSlice);
        }
    }

    private final String rawFormatString;
    private final Object keywordArgs;
    private final List<Object> positionalArgs;

    // Contents are literal string, index into positional arg, or
    // treeDataKey, which is list of objects as path into treeData.
    // Each part of treeDataKey in turn consists of JSON property name, or index into JSON array.
    private final List<PartDescriptor> parsedFormatString;

    public LogMessageTemplate(String formatString, Object keywordArgs,
            List<Object> positionalArgs) {
        this(formatString, new LogMessageTemplateParser(formatString).parse(),
            keywordArgs, positionalArgs);
    }

    public LogMessageTemplate(String rawFormatString, List<PartDescriptor> parsedFormatString,
            Object keywordArgs, List<Object> positionalArgs) {
        this.rawFormatString = rawFormatString;
        this.parsedFormatString = parsedFormatString;
        this.keywordArgs = keywordArgs;
        this.positionalArgs = positionalArgs;
    }

    public List<PartDescriptor> getParsedFormatString() {
        return parsedFormatString;
    }

    @Override
    public String toString() {
        List<Object> formatArgs = new ArrayList<>();
        String genericFormatString = generateFormatString(formatArgs, false);
        String msg = String.format(genericFormatString, formatArgs.toArray());
        return msg;
    }

    public Object toStructuredLogRecord() {
        return toStructuredLogRecord(keywordArgs);
    }

    public Unstructured toUnstructuredLogRecord() {
        List<Object> formatArgs = new ArrayList<>();
        String formatString = generateFormatString(formatArgs, true);
        return new Unstructured(formatString, formatArgs);
    }

    private Object toStructuredLogRecord(Object treeDataSlice) {
        return new Structured(treeDataSlice);
    }

    protected abstract String serializeData(Object treeDataSlice);/* {
        try {
            Class<?> gsonCls = Class.forName("com.google.gson.Gson");
            Object gsonInstance = gsonCls.newInstance();
            Method method = gsonCls.getMethod("toJson", Object.class);
            return (String)method.invoke(gsonInstance, treeDataSlice);
        }
        catch (Exception ex) {
            if (ex instanceof RuntimeException) {
                throw (RuntimeException)ex;
            }
            throw new RuntimeException(ex);
        }
    }*/

    protected String escapeLiteral(String literal, boolean forLogger) {
        return literal.replace("%", "%%");
    }

    protected String generatePositionIndicator(int position, boolean forLogger) {
        return "%s";
    }

    Object getPositionalArg(List<Object> args, PartDescriptor part) {
        if (args == null) {
            return handleNonExistentPositionalArg(part);
        }
        // Support Python style indexing.
        int index = part.positionalArgIndex;
        if (Math.abs(index) >= args.size()) {
            return handleNonExistentPositionalArg(part);
        }
        if (index < 0) {
            index += args.size();
        }
        return args.get(index);
    }

    @SuppressWarnings("unchecked")
    Object getTreeDataSlice(Object treeData, PartDescriptor part) {
        Object treeDataSlice = treeData;
        for (int i = 0; i < part.treeDataKey.size(); i++) {
            if (treeDataSlice == null) {
                return handleNonExistentTreeDataSlice(part, i);
            }
            Object keyPart = part.treeDataKey.get(i);
            if (keyPart instanceof Integer) {
                int index = (Integer) keyPart;
                if (!(treeDataSlice instanceof List)) {
                    treeDataSlice = getTreeDataListItem(treeDataSlice, index);
                    continue;
                }
                List<Object> jsonArray = (List<Object>) treeDataSlice;
                // Support Python style indexing.
                if (Math.abs(index) >= jsonArray.size()) {
                    return handleNonExistentTreeDataSlice(part, i);
                }
                if (index < 0) {
                    index += jsonArray.size();
                }
                treeDataSlice = jsonArray.get(index);
            }
            else {
                if (!(treeDataSlice instanceof Map)) {
                    treeDataSlice = getTreeDataPropertyValue(treeDataSlice, keyPart);
                    continue;
                }
                Map<Object, Object> jsonObject = (Map<Object, Object>) treeDataSlice;
                if (!jsonObject.containsKey(keyPart)) {
                    return handleNonExistentTreeDataSlice(part, i);
                }
                treeDataSlice = jsonObject.get(keyPart);
            }
        }
        return treeDataSlice;
    }

    protected Object handleNonExistentTreeDataSlice(PartDescriptor part, int nonExistentIndex) {
        return rawFormatString.substring(part.startPos, part.endPos);
    }

    protected Object handleNonExistentPositionalArg(PartDescriptor part) {
        return rawFormatString.substring(part.startPos, part.endPos);
    }

    protected Object getTreeDataListItem(Object treeData, int index) {
        return null;
    }

    protected Object getTreeDataPropertyValue(Object treeData, Object propertyName) {
        try {
            Class<?> propAccessorType = Class.forName("org.apache.commons.beanutils.PropertyUtils");
            Method propAccessor = propAccessorType.getMethod("getSimpleProperty", Object.class,
                String.class);
            return propAccessor.invoke(null, treeData, propertyName);
        }
        catch (Exception ex) {
            return null;
        }
    }

    private String generateFormatString(List<Object> formatArgsReceiver, boolean forLogger) {
        StringBuilder logFormat = new StringBuilder();
        int uniqueIndex = 0;
        for (PartDescriptor part : parsedFormatString) {
            if (part.literalSection != null) {
                logFormat.append(escapeLiteral(part.literalSection, forLogger));
            }
            else {
                logFormat.append(generatePositionIndicator(uniqueIndex++, forLogger));
                Object formatArg;
                if (part.treeDataKey != null) {
                    formatArg = getTreeDataSlice(keywordArgs, part);
                }
                else {
                    formatArg = getPositionalArg(positionalArgs, part);
                }
                if (part.serialize) {
                    formatArgsReceiver.add(toStructuredLogRecord(formatArg));
                }
                else {
                    formatArgsReceiver.add(formatArg);
                }
            }
        }
        return logFormat.toString();
    }
}

File: csharp/ShrewdEvolver/LogMessageTemplate.cs
using System;
using System.Collections.Generic;
using System.Text;
using static AaronicSubstances.ShrewdEvolver.LogMessageTemplateParser;

namespace AaronicSubstances.ShrewdEvolver
{
    public class LogMessageTemplate
    {
        public class Unstructured
        {
            public Unstructured(string formatString, IList<object> formatArgs)
            {
                FormatString = formatString;
                FormatArgs = formatArgs;
            }
            public string FormatString { get; }
            public IList<object> FormatArgs { get; }
        }

        public class Structured
        {
            private readonly LogMessageTemplate _logRecord;
            private readonly object _treeDataSlice;
            public Structured(LogMessageTemplate logRecord, object treeDataSlice)
            {
                _logRecord = logRecord;
                _treeDataSlice = treeDataSlice;
            }
            public override string ToString()
            {
                return _logRecord.SerializeData(_treeDataSlice);
            }
        }

        private readonly object _keywordArgs;
        private readonly IList<object> _positionalArgs;

        // Contents are literal string, index into positional arg, or
        // treeDataKey, which is list of objects as path into treeData.
        // Each part of treeDataKey in turn consists of JSON property name, or index into JSON array.
        private readonly IList<PartDescriptor> _parsedFormatString;

        public LogMessageTemplate(string formatString, object keywordArgs,
                IList<object> positionalArgs)
        {
            _keywordArgs = keywordArgs;
            _positionalArgs = positionalArgs;
            _parsedFormatString = ParseFormatString(formatString);
        }

        public override string ToString()
        {
            var formatArgs = new List<object>();
            string genericFormatString = GenerateFormatString(formatArgs, false);
            string msg = string.Format(genericFormatString, formatArgs.ToArray());
            return msg;
        }

        public object ToStructuredLogRecord()
        {
            return ToStructuredLogRecord(_keywordArgs);
        }

        public Unstructured ToUnstructuredLogRecord()
        {
            var formatArgs = new List<object>();
            string formatString = GenerateFormatString(formatArgs, true);
            return new Unstructured(formatString, formatArgs);
        }

        protected virtual object ToStructuredLogRecord(object treeDataSlice)
        {
            return new Structured(this, treeDataSlice);
        }

        protected virtual string SerializeData(object treeDataSlice)
        {
            throw new NotImplementedException();
        }

        protected virtual string EscapeLiteral(string literal, bool forLogger)
        {
            return literal.Replace("{", "{{").Replace("}", "}}");
        }

        protected virtual string GeneratePositionIndicator(int position, bool forLogger)
        {
            return "{" + position + "}";
        }

        protected virtual object GetPositionalArg(IList<object> args, int index)
        {
            // Support Python style indexing.
            if (Math.Abs(index) >= args.Count)
            {
                return HandleNonExistentPositionalArg(index);
            }
            if (index < 0)
            {
                index += args.Count;
            }
            return args[index];
        }

        protected virtual object GetTreeDataSlice(object treeData, IList<object> treeDataKey)
        {
            object treeDataSlice = treeData;
            foreach (object keyPart in treeDataKey)
            {
                if (treeDataSlice == null)
                {
                    break;
                }
                if (keyPart is int) 
                {
                    var index = (int)keyPart;
                    if (!(treeDataSlice is IList<object>))
                    {
                        treeDataSlice = GetTreeDataListItem(treeDataSlice, index);
                        continue;
                    }
                    var jsonArray = (IList<object>)treeDataSlice;
                    // Support Python style indexing.
                    if (Math.Abs(index) >= jsonArray.Count)
                    {
                        return HandleNonExistentTreeDataSlice(treeDataKey);
                    }
                    if (index < 0)
                    {
                        index += jsonArray.Count;
                    }
                    treeDataSlice = jsonArray[index];
                }
                else
                {
                    var key = (string)keyPart;
                    if (!(treeDataSlice is IDictionary<string, object>))
                    {
                        treeDataSlice = GetTreeDataPropertyValue(treeDataSlice, keyPart);
                        continue;
                    }
                    var jsonObject = (IDictionary<string, object>)treeDataSlice;
                    if (!jsonObject.ContainsKey(key))
                    {
                        return HandleNonExistentTreeDataSlice(treeDataKey);
                    }
                    treeDataSlice = jsonObject[key];
                }
            }
            return treeDataSlice;
        }

        protected virtual object HandleNonExistentPositionalArg(int index)
        {
            return null;
        }

        protected virtual object HandleNonExistentTreeDataSlice(IList<object> treeDataKey)
        {
            return null;
        }

        protected virtual object GetTreeDataListItem(object treeData, int index)
        {
            return null;
        }

        protected virtual object GetTreeDataPropertyValue(object treeData, object propertyName)
        {
            try
            {
                var propHandle = treeData.GetType().GetProperty((string) propertyName);
                return propHandle.GetValue(treeData);
            }
            catch (Exception) 
            {
                return null;
            }
        }

        internal static IList<PartDescriptor> ParseFormatString(string formatString)
        {
            return new LogMessageTemplateParser(formatString).Parse();
        }

        internal IList<PartDescriptor> GetParsedFormatString()
        {
            return _parsedFormatString;
        }

        private string GenerateFormatString(IList<object> formatArgsReceiver, bool forLogger)
        {
            var logFormat = new StringBuilder();
            int uniqueIndex = 0;
            foreach (PartDescriptor part in _parsedFormatString)
            {
                if (part.treeDataKey != null)
                {
                    logFormat.Append(GeneratePositionIndicator(uniqueIndex++, forLogger));
                    object treeDataSlice = GetTreeDataSlice(_keywordArgs, part.treeDataKey);
                    if (part.serialize)
                    {
                        formatArgsReceiver.Add(ToStructuredLogRecord(treeDataSlice));
                    }
                    else
                    {
                        formatArgsReceiver.Add(treeDataSlice);
                    }
                }
                else if (part.literalSection != null)
                {
                    logFormat.Append(EscapeLiteral(part.literalSection, forLogger));
                }
                else
                {
                    logFormat.Append(GeneratePositionIndicator(uniqueIndex++, forLogger));
                    object item = GetPositionalArg(_positionalArgs, part.positionalArgIndex);
                    if (part.serialize)
                    {
                        formatArgsReceiver.Add(ToStructuredLogRecord(item));
                    }
                    else
                    {
                        formatArgsReceiver.Add(item);
                    }
                }
            }
            return logFormat.ToString();
        }
    }
}
